### DISCUSSION TITLE
Help configuring Mapstruct (Annotation processor)

### DISCUSSION THREAD

**Discussion by bavo** [mfussenegger/nvim-jdtls#718] [Q&A]
I'm trying to include the mapstruct annotation processor into my vim-jdtls configuration. 
But I don't seem to get it to work. 

It's this supported?
Am I doing something wrong? 
Does Lombok use the same method for it's code generation? 

Annotation processor:  https://docs.oracle.com/javase/6/docs/technotes/guides/apt/index.html

Mapstruct documentation: https://mapstruct.org/documentation/stable/reference/html/#introduction

**Comment by fromSmolsoft** [comment]
Mapstruct works fain in my gradle project opened with nvim-jdtls just by adding 
```groovy
//...
ext {
    mapstructVersion = '1.6.3'
}
//...
dependencies {
    //...
    implementation "org.mapstruct:mapstruct:${mapstructVersion}"
    annotationProcessor "org.mapstruct:mapstruct-processor:${mapstructVersion}"
}
```

I can't think of any specific config for mapstruct in jdtls. 
Then again i do mass load all the bundles aside from "com.microsoft.java.test.runner-jar-with-dependencies.jar", "jacocoagent.jar". 

_Note: For lombok i use gradle lombok plugin and in nvim-jdtls config in cmd i have `"--jvm-arg=-javaagent:" .. path/to/lombok.jar`  -> has to lead to jdtls/lombok.jar plugin wherever you have it._

**Comment by mukuldaroch** [comment]
I am having similar issues with both gradle and with maven 
this is my nvim-jdtls config file 
-------------
```lua
return {
    "mfussenegger/nvim-jdtls",
    ft = { "java" },
    config = function()
        local home = os.getenv("HOME")
        local workspace_path = home .. "/.local/share/nvim/jdtls-workspace/"
        local project_name = vim.fn.fnamemodify(vim.fn.getcwd(), ":p:h:t")
        local workspace_dir = workspace_path .. project_name

        local status, jdtls = pcall(require, "jdtls")
        if not status then
            return
        end

        local extendedClientCapabilities = jdtls.extendedClientCapabilities

        -- =========================
        --  Full Lombok + JDTLS Config
        -- =========================
        local config = {
            cmd = {
                "java",
                "-Declipse.application=org.eclipse.jdt.ls.core.id1",
                "-Dosgi.bundles.defaultStartLevel=4",
                "-Declipse.product=org.eclipse.jdt.ls.core.product",
                "-Dlog.protocol=true",
                "-Dlog.level=ALL",
                "-Xmx2g",
                "--add-modules=ALL-SYSTEM",
                "--add-opens", "java.base/java.util=ALL-UNNAMED",
                "--add-opens", "java.base/java.lang=ALL-UNNAMED",

                --  Lombok Agent
                "-javaagent:" .. home .. "/.local/share/nvim/mason/packages/jdtls/lombok.jar",

                --  JDTLS launcher
                "-jar",
                vim.fn.glob(home .. "/.local/share/nvim/mason/packages/jdtls/plugins/org.eclipse.equinox.launcher_*.jar"),

                -- Linux config
                "-configuration",
                home .. "/.local/share/nvim/mason/packages/jdtls/config_linux",

                -- Workspace per project
                "-data",
                workspace_dir,
            },

            -- Root dir detection (Gradle)
            root_dir = require("jdtls.setup").find_root({
                ".git",
                "gradlew",
                "build.gradle",
            }),

            -- =========================
            -- ⚙️ Java Language Settings
            -- =========================
            settings = {
                java = {
                    signatureHelp = { enabled = true },
                    extendedClientCapabilities = extendedClientCapabilities,
                    referencesCodeLens = { enabled = true },
                    references = { includeDecompiledSources = true },
                    inlayHints = { parameterNames = { enabled = "all" } },
                    format = { enabled = false }, -- disable LSP formatting
                    projectSources = {
                        vim.fn.getcwd() .. "/src/main/java",
                        vim.fn.getcwd() .. "/build/generated/sources/annotationProcessor/java/main",
                    },
                },
            },

            init_options = {
                bundles = {},
            },
        }

        -- =========================
        -- Start or Attach JDTLS
        -- =========================
        jdtls.start_or_attach(config)
    end,
}
```
------------------------------------------------------------------------------------------------------------------------------------------------
iam facing similar issues if i use maven the annotations don't process and the lsp gives error on certains functions like getters and setters or mapstruct interface 
and with gradle  lsp give a ton of errors on Imports and functions genrated by getter and setter annotations etc 

This is a example file of my build.gradle
------------------------------------------------------------------------------------------------------------------------------------------------
```
plugins {
    id 'java'
    id 'idea'
    id 'org.springframework.boot' version '3.5.6'
    id 'io.spring.dependency-management' version '1.1.3'
}

group = 'com.daroch'
version = '0.0.1-SNAPSHOT'

java {
    sourceCompatibility = JavaVersion.VERSION_17
    targetCompatibility = JavaVersion.VERSION_17
}

repositories {
    mavenCentral()
}

ext {
    mapstructVersion = '1.6.3'
    lombokVersion = '1.18.32'
}

dependencies {
    // Spring Boot
    implementation 'org.springframework.boot:spring-boot-starter-data-jpa'
    implementation 'org.springframework.boot:spring-boot-starter-web'
    implementation 'org.springframework.boot:spring-boot-starter-security'
    implementation 'org.springframework.boot:spring-boot-starter-oauth2-resource-server'
    implementation 'org.springframework.boot:spring-boot-starter-validation'

    // MapStruct + Lombok
    compileOnly "org.projectlombok:lombok:${lombokVersion}"
    annotationProcessor "org.projectlombok:lombok:${lombokVersion}"
    implementation "org.mapstruct:mapstruct:${mapstructVersion}"
    annotationProcessor "org.mapstruct:mapstruct-processor:${mapstructVersion}"
    annotationProcessor "org.projectlombok:lombok-mapstruct-binding:0.2.0"

    // Databases
    runtimeOnly 'com.h2database:h2'
    runtimeOnly 'org.postgresql:postgresql'

    // Testing
    testImplementation 'org.springframework.boot:spring-boot-starter-test'
    testImplementation 'org.springframework.security:spring-security-test'
}

tasks.withType(Test).configureEach {
    // ✅ Gradle 8.8+ syntax — works for Gradle 9.1
    failOnNoDiscoveredTests = false
}

tasks.withType(JavaCompile).configureEach {
    options.annotationProcessorPath = configurations.annotationProcessor
    options.compilerArgs += [
        "-Amapstruct.defaultComponentModel=spring",
        "-Amapstruct.verbose=true" // <- enable verbose mode for debugging
    ]
}
sourceSets.main.java.srcDirs += 'build/generated/sources/annotationProcessor/java/main'

bootJar {
    enabled = true
}
```

**Reply by fromSmolsoft** [reply]
@mukuldaroch  I would start by running the project with gradle / maven from command line. If it starts/ works it's jdtls config issue. 

if it is jdtls:
Do you by any chance use (automatic) `ls` setup via `mason-lspconfig.nvim` potentially with `nvim-lspconfig`? It can start it's own version of jdtls aside nvim-jdtls. AFAIK pre-configured jdtls is quite barebone. Aside from :LspInfo you can see that code actions are doubled.

**Reply by mukuldaroch** [reply]
Thanks it was indeed jdtls file config issue i tried your nvim config files and it kind of worked then i configured my own config files and added your configuration for jdtls and it worked 

`By the way i have one doubt ` when i open any java file the gradle automatically start working after jdtls is that normal ? 
and even after i exit the nvim the gradle daemon continues running  

```bash
ps aux | grep java

daroch     58936  5.3  4.1 5631196 672464 ?      Ssl  14:43   1:32 /usr/lib/jvm/java-21-openjdk/bin/java --add-opens=java.base/java.lang=ALL-UNNAMED --add-opens=java.base/java.lang.invoke=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.prefs/java.util.prefs=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.api=ALL-UNNAMED --add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED --add-opens=java.base/java.util=ALL-UNNAMED --add-opens=java.prefs/java.util.prefs=ALL-UNNAMED --add-opens=java.base/java.nio.charset=ALL-UNNAMED --add-opens=java.base/java.net=ALL-UNNAMED --add-opens=java.base/java.util.concurrent=ALL-UNNAMED --add-opens=java.base/java.util.concurrent.atomic=ALL-UNNAMED --add-opens=java.xml/javax.xml.namespace=ALL-UNNAMED --add-opens=java.base/java.time=ALL-UNNAMED -XX:MaxMetaspaceSize=384m -XX:+HeapDumpOnOutOfMemoryError -Xms256m -Xmx512m -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -cp /home/daroch/.gradle/wrapper/dists/gradle-9.1.0-bin/9agqghryom9wkf8r80qlhnts3/gradle-9.1.0/lib/gradle-daemon-main-9.1.0.jar -javaagent:/home/daroch/.gradle/wrapper/dists/gradle-9.1.0-bin/9agqghryom9wkf8r80qlhnts3/gradle-9.1.0/lib/agents/gradle-instrumentation-agent-9.1.0.jar org.gradle.launcher.daemon.bootstrap.GradleDaemon 9.1.0

daroch     69219 13.2  7.8 10141416 1270600 ?    Ssl  15:03   1:11 -Declipse.application=org.eclipse.jdt.ls.core.id1 -Dosgi.bundles.defaultStartLevel=4 -Declipse.product=org.eclipse.jdt.ls.core.product -Dosgi.checkConfiguration=true -Dosgi.sharedConfiguration.area=/home/daroch/.local/share/nvim/mason/packages/jdtls/config_linux -Dosgi.sharedConfiguration.area.readOnly=true -Dosgi.configuration.cascaded=true -Xms1G --add-modules=ALL-SYSTEM --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED -javaagent:/home/daroch/.local/share/nvim/mason/packages/jdtls/lombok.jar -jar /home/daroch/.local/share/nvim/mason/packages/jdtls/plugins/org.eclipse.equinox.launcher_1.7.0.v20250519-0528.jar -data /home/daroch/.cache/jdtls/workspace/tickets -Declipse.application=org.eclipse.jdt.ls.core.id1 -Dosgi.bundles.defaultStartLevel=4 -Declipse.product=org.eclipse.jdt.ls.core.product -Dlog.protocol=true -Dlog.level=ALL -Xmx1G --add-modules=ALL-SYSTEM --add-opens java.base/java.util=ALL-UNNAMED --add-opens java.base/java.lang=ALL-UNNAMED -jar /home/daroch/.local/share/nvim/mason/packages/jdtls/plugins/org.eclipse.equinox.launcher_1.7.0.v20250519-0528.jar -configuration /home/daroch/.local/share/nvim/mason/packages/jdtls/config_linux

daroch     72297  0.0  0.0   6604  4284 pts/3    S+   15:12   0:00 grep java
```
thanks for your guidance and support

**Reply by fromSmolsoft** [reply]
@mukuldaroch 
> i tried your nvim config files 

That nvim config in my repo is bit old but I'm glad it still works somewhat.  I have to update it when i get free time. 
 
> when i open any java file the gradle automatically start working

No it is not normal when opening nay java file.  Dradle daemon is fired only when **gradle** project is opened .  But for gradle projects it's normal  AFAIK [reported "issue" of this very same behavior](https://github.com/mfussenegger/nvim-jdtls/discussions/399)
In short Jdtls uses gradle for some features and gradle daemon itself by default wait idle for up to 3 hrs. It's not jdtls misconfiguration.  [gradle doc](https://docs.gradle.org/current/userguide/gradle_daemon.html#sec:stopping_an_existing_daemon)
